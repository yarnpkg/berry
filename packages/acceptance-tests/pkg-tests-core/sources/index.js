const exec = require('./utils/exec');
const fs = require('./utils/fs');
const misc = require('./utils/misc');
const tests = require('./utils/tests');

module.exports = {exec, fs, misc, tests};
