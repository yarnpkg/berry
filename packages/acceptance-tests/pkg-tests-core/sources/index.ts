import './utils/makeTemporaryEnv';
import './utils/makeTemporaryMonorepoEnv';
import * as exec  from './utils/exec';
import * as fs    from './utils/fs';
import * as misc  from './utils/misc';
import * as tests from './utils/tests';
import * as yarn  from './utils/yarn';

export {exec, fs, misc, tests, yarn};
