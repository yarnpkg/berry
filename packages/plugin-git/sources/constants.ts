export const GIT_REGEXP = /^(git:|git\+.+:|ssh:|https?:.+\.git#.+|https?:.+\.git)/;
