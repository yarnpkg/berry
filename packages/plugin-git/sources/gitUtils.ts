import {Configuration, Locator, execUtils, structUtils, httpUtils, semverUtils} from '@yarnpkg/core';
import {npath, xfs}                                                             from '@yarnpkg/fslib';
import GitUrlParse                                                              from 'git-url-parse';
import querystring                                                              from 'querystring';
import semver                                                                   from 'semver';
import urlLib                                                                   from 'url';

function makeGitEnvironment() {
  return {
    ...process.env,
    // An option passed to SSH by Git to prevent SSH from asking for data (which would cause installs to hang when the SSH keys are missing)
    GIT_SSH_COMMAND: `ssh -o BatchMode=yes`,
  };
}

const gitPatterns = [
  /^ssh:/,
  /^git(?:\+[^:]+)?:/,

  // `git+` is optional, `.git` is required
  /^(?:git\+)?https?:[^#]+\/[^#]+(?:\.git)(?:#.*)?$/,

  /^git@[^#]+\/[^#]+\.git(?:#.*)?$/,

  /^(?:github:|https:\/\/github\.com\/)?(?!\.{1,2}\/)([a-zA-Z._0-9-]+)\/(?!\.{1,2}(?:#|$))([a-zA-Z._0-9-]+?)(?:\.git)?(?:#.*)?$/,
  // GitHub `/tarball/` URLs
  /^https:\/\/github\.com\/(?!\.{1,2}\/)([a-zA-Z0-9._-]+)\/(?!\.{1,2}(?:#|$))([a-zA-Z0-9._-]+?)\/tarball\/(.+)?$/,
];

export enum TreeishProtocols {
  Commit = `commit`,
  Head = `head`,
  Tag = `tag`,
  Semver = `semver`,
}

/**
 * Determines whether a given url is a valid github git url via regex
 */
export function isGitUrl(url: string): boolean {
  return url ? gitPatterns.some(pattern => !!url.match(pattern)) : false;
}

export type RepoUrlParts = {
  repo: string;
  treeish: {
    protocol: TreeishProtocols | string | null,
    request: string,
  },
  extra: {
    [key: string]: string,
  },
};

export function splitRepoUrl(url: string): RepoUrlParts {
  url = normalizeRepoUrl(url);

  const hashIndex = url.indexOf(`#`);
  if (hashIndex === -1) {
    return {
      repo: url,
      treeish: {
        protocol: TreeishProtocols.Head,
        request: `HEAD`,
      },
      extra: {},
    };
  }

  const repo = url.slice(0, hashIndex);
  const subsequent = url.slice(hashIndex + 1);

  // New-style: "#commit=abcdef&workspace=foobar"
  if (subsequent.match(/^[a-z]+=/)) {
    const extra = querystring.parse(subsequent);

    for (const [key, value] of Object.entries(extra))
      if (typeof value !== `string`)
        throw new Error(`Assertion failed: The ${key} parameter must be a literal string`);

    const requestedProtocol = Object.values(TreeishProtocols).find(protocol => {
      return Object.prototype.hasOwnProperty.call(extra, protocol);
    });

    let protocol: TreeishProtocols;
    let request: string;

    if (typeof requestedProtocol !== `undefined`) {
      protocol = requestedProtocol;
      request = extra[requestedProtocol]! as string;
    } else {
      protocol = TreeishProtocols.Head;
      request = `HEAD`;
    }

    for (const key of Object.values(TreeishProtocols))
      delete extra[key];

    return {
      repo,
      treeish: {protocol, request},
      extra: extra as {
        [key: string]: string,
      },
    };
  } else {
    // Old-style: "#commit:abcdef" or "#abcdef"
    const colonIndex = subsequent.indexOf(`:`);

    let protocol: string | null;
    let request: string;

    if (colonIndex === -1) {
      protocol = null;
      request = subsequent;
    } else {
      protocol = subsequent.slice(0, colonIndex);
      request = subsequent.slice(colonIndex + 1);
    }

    return {
      repo,
      treeish: {protocol, request},
      extra: {},
    };
  }
}

export function normalizeRepoUrl(url: string, {git = false}: {git?: boolean} = {}) {
  // "git+https://" isn't an actual Git protocol. It's just a way to
  // disambiguate that this URL points to a Git repository.
  url = url.replace(/^git\+https:/, `https:`);

  // We support this as an alias to GitHub repositories
  url = url.replace(/^(?:github:|https:\/\/github\.com\/)?(?!\.{1,2}\/)([a-zA-Z0-9._-]+)\/(?!\.{1,2}(?:#|$))([a-zA-Z0-9._-]+?)(?:\.git)?(#.*)?$/, `https://github.com/$1/$2.git$3`);

  // We support GitHub `/tarball/` URLs
  url = url.replace(/^https:\/\/github\.com\/(?!\.{1,2}\/)([a-zA-Z0-9._-]+)\/(?!\.{1,2}(?:#|$))([a-zA-Z0-9._-]+?)\/tarball\/(.+)?$/, `https://github.com/$1/$2.git#$3`);

  if (git) {
    // The `git+` prefix doesn't mean anything at all for Git
    url = url.replace(/^git\+([^:]+):/, `$1:`);

    // The `ssh://` prefix should be removed because so URLs won't work in Git:
    //   ssh://git@github.com:yarnpkg/berry.git
    //   git@github.com/yarnpkg/berry.git
    // Git only allows:
    //   git@github.com:yarnpkg/berry.git (no ssh)
    //   ssh://git@github.com/yarnpkg/berry.git (no colon)
    // So we should cut `ssh://`, but only in URLs that contain colon after the hostname

    let parsedUrl: urlLib.UrlWithStringQuery | null;
    try {
      parsedUrl = urlLib.parse(url);
    } catch {
      parsedUrl = null;
    }

    if (parsedUrl && parsedUrl.protocol === `ssh:` && parsedUrl.path?.startsWith(`/:`)) {
      url = url.replace(/^ssh:\/\//, ``);
    }
  }

  return url;
}

export function normalizeLocator(locator: Locator) {
  return structUtils.makeLocator(locator, normalizeRepoUrl(locator.reference));
}

export async function lsRemote(repo: string, configuration: Configuration) {
  const normalizedRepoUrl = normalizeRepoUrl(repo, {git: true});

  const networkSettings = httpUtils.getNetworkSettings(`https://${GitUrlParse(normalizedRepoUrl).resource}`, {configuration});
  if (!networkSettings.enableNetwork)
    throw new Error(`Request to '${normalizedRepoUrl}' has been blocked because of your configuration settings`);

  let res: {stdout: string};
  try {
    res = await execUtils.execvp(`git`, [`ls-remote`, normalizedRepoUrl], {
      cwd: configuration.startingCwd,
      env: makeGitEnvironment(),
      strict: true,
    });
  } catch (error) {
    error.message = `Listing the refs for ${repo} failed`;
    throw error;
  }

  const refs = new Map();

  const matcher = /^([a-f0-9]{40})\t([^\n]+)/gm;
  let match;

  while ((match = matcher.exec(res.stdout)) !== null)
    refs.set(match[2], match[1]);

  return refs;
}

export async function resolveUrl(url: string, configuration: Configuration) {
  const {repo, treeish: {protocol, request}, extra} = splitRepoUrl(url);
  const refs = await lsRemote(repo, configuration);

  const resolve = (protocol: TreeishProtocols | string | null, request: string): string => {
    switch (protocol) {
      case TreeishProtocols.Commit: {
        if (!request.match(/^[a-f0-9]{40}$/))
          throw new Error(`Invalid commit hash`);

        return querystring.stringify({
          ...extra,
          commit: request,
        });
      }

      case TreeishProtocols.Head: {
        const head = refs.get(request === `HEAD` ? request : `refs/heads/${request}`);
        if (typeof head === `undefined`)
          throw new Error(`Unknown head ("${request}")`);

        return querystring.stringify({
          ...extra,
          commit: head,
        });
      }

      case TreeishProtocols.Tag: {
        const tag = refs.get(`refs/tags/${request}`);
        if (typeof tag === `undefined`)
          throw new Error(`Unknown tag ("${request}")`);

        return querystring.stringify({
          ...extra,
          commit: tag,
        });
      }

      case TreeishProtocols.Semver: {
        const validRange = semverUtils.validRange(request);
        if (!validRange)
          throw new Error(`Invalid range ("${request}")`);

        const semverTags = new Map([...refs.entries()].filter(([ref]) => {
          return ref.startsWith(`refs/tags/`);
        }).map<[semver.SemVer | null, string]>(([ref, hash]) => {
          return [semver.parse(ref.slice(10)), hash];
        }).filter((entry): entry is [semver.SemVer, string] => {
          return entry[0] !== null;
        }));

        const bestVersion = semver.maxSatisfying([...semverTags.keys()], validRange);
        if (bestVersion === null)
          throw new Error(`No matching range ("${request}")`);

        return querystring.stringify({
          ...extra,
          commit: semverTags.get(bestVersion),
        });
      }

      case null: {
        let result: string | null;

        if ((result = tryResolve(TreeishProtocols.Commit, request)) !== null)
          return result;
        if ((result = tryResolve(TreeishProtocols.Tag, request)) !== null)
          return result;
        if ((result = tryResolve(TreeishProtocols.Head, request)) !== null)
          return result;

        if (request.match(/^[a-f0-9]+$/)) {
          throw new Error(`Couldn't resolve "${request}" as either a commit, a tag, or a head - if a commit, use the 40-characters commit hash`);
        } else {
          throw new Error(`Couldn't resolve "${request}" as either a commit, a tag, or a head`);
        }
      }

      default: {
        throw new Error(`Invalid Git resolution protocol ("${protocol}")`);
      }
    }
  };

  const tryResolve = (protocol: TreeishProtocols | string | null, request: string): string | null => {
    try {
      return resolve(protocol, request);
    } catch (err) {
      return null;
    }
  };

  return `${repo}#${resolve(protocol, request)}`;
}

export async function clone(url: string, configuration: Configuration) {
  return await configuration.getLimit(`cloneConcurrency`)(async () => {
    const {repo, treeish: {protocol, request}} = splitRepoUrl(url);
    if (protocol !== `commit`)
      throw new Error(`Invalid treeish protocol when cloning`);

    const normalizedRepoUrl = normalizeRepoUrl(repo, {git: true});
    if (httpUtils.getNetworkSettings(`https://${GitUrlParse(normalizedRepoUrl).resource}`, {configuration}).enableNetwork === false)
      throw new Error(`Request to '${normalizedRepoUrl}' has been blocked because of your configuration settings`);

    const directory = await xfs.mktempPromise();
    const execOpts = {cwd: directory, env: makeGitEnvironment(), strict: true};

    try {
      await execUtils.execvp(`git`, [`clone`, `-c core.autocrlf=false`, normalizedRepoUrl, npath.fromPortablePath(directory)], execOpts);
      await execUtils.execvp(`git`, [`checkout`, `${request}`], execOpts);
    } catch (error) {
      error.message = `Repository clone failed: ${error.message}`;
      throw error;
    }

    return directory;
  });
}
