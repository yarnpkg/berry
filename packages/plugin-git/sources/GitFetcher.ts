import {Fetcher, FetchOptions, MinimalFetchOptions, FetchResult} from '@yarnpkg/core';
import {Locator, MessageName}                                    from '@yarnpkg/core';
import {miscUtils, scriptUtils, structUtils, tgzUtils}           from '@yarnpkg/core';
import {PortablePath, ppath, xfs}                                from '@yarnpkg/fslib';

import * as gitUtils                                             from './gitUtils';
import {Hooks}                                                   from './index';

export class GitFetcher implements Fetcher {
  supports(locator: Locator, opts: MinimalFetchOptions) {
    return gitUtils.isGitUrl(locator.reference);
  }

  getLocalPath(locator: Locator, opts: FetchOptions) {
    return null;
  }

  async fetch(locator: Locator, opts: FetchOptions) {
    const expectedChecksum = opts.checksums.get(locator.locatorHash) || null;

    const result = await this.downloadHosted(locator, opts);
    if (result !== null)
      return result;

    const [packageFs, releaseFs, checksum] = await opts.cache.fetchPackageFromCache(
      locator,
      expectedChecksum,
      async () => {
        opts.report.reportInfoOnce(MessageName.FETCH_NOT_CACHED, `${structUtils.prettyLocator(opts.project.configuration, locator)} can't be found in the cache and will be fetched from the remote repository`);
        return await this.cloneFromRemote(locator, opts);
      },
    );

    return {
      packageFs,
      releaseFs,
      prefixPath: structUtils.getIdentVendorPath(locator),
      checksum,
    };
  }

  async downloadHosted(locator: Locator, opts: FetchOptions) {
    return opts.project.configuration.reduceHook((hooks: Hooks) => {
      return hooks.fetchHostedRepository;
    }, null as FetchResult | null, locator, opts);
  }

  async cloneFromRemote(locator: Locator, opts: FetchOptions) {
    const cloneTarget = await gitUtils.clone(locator.reference, opts.project.configuration);

    const packagePath = ppath.join(cloneTarget, `package.tgz` as PortablePath);
    await scriptUtils.prepareExternalProject(cloneTarget, packagePath, {
      configuration: opts.project.configuration,
      report: opts.report,
    });

    const sourceBuffer = await xfs.readFilePromise(packagePath);

    return await miscUtils.releaseAfterUseAsync(async () => {
      return await tgzUtils.convertToZip(sourceBuffer, {
        stripComponents: 1,
        prefixPath: structUtils.getIdentVendorPath(locator),
      });
    });
  }
}
