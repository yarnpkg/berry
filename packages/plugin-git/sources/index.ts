import {FetchOptions, FetchResult, Locator, Plugin, SettingsType} from '@yarnpkg/core';

import {GitFetcher}                                               from './GitFetcher';
import {GitResolver}                                              from './GitResolver';
import * as gitUtils                                              from './gitUtils';

export interface Hooks {
  fetchHostedRepository?: (
    current: FetchResult | null,
    locator: Locator,
    opts: FetchOptions,
  ) => Promise<FetchResult | null>,
}

declare module '@yarnpkg/core' {
  interface ConfigurationValueMap {
    cloneConcurrency: number;
  }
}

const plugin: Plugin = {
  configuration: {
    cloneConcurrency: {
      description: `Maximal number of concurrent clones`,
      type: SettingsType.NUMBER,
      default: 2,
    },
  },
  fetchers: [
    GitFetcher,
  ],
  resolvers: [
    GitResolver,
  ],
};

export {gitUtils};

// eslint-disable-next-line arca/no-default-export
export default plugin;
