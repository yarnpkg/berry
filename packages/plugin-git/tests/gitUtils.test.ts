import * as gitUtils from '../sources/gitUtils';

const VALID_PATTERNS = [
  `GitHubOrg/foo-bar.js`,
  `GitHubOrg/foo2bar.js`,
  `GitHubOrg/foo-bar.js#hash`,
  `GitHubOrg/foo-bar.js#commit:hash`,
  `GitHubOrg/foo-bar.js#commit=hash`,
  `GitHubOrg/foo-bar.js#commit=hash&workspace=foo`,
  `GitHubOrg/foo-bar.js#tag=hello`,
  `GitHubOrg/foo-bar.js#workspace=foo`,
  `github:GitHubOrg/foo-bar.js`,
  `github:GitHubOrg/foo-bar.js#hash`,
  `https://github.com/TooTallNate/util-deprecate.git#v1.0.0`,
  `https://github.com/TooTallNate/util-deprecate.git#semver:v1.0.0`,
  `https://github.com/TooTallNate/util-deprecate.git#master`,
  `https://github.com/TooTallNate/util-deprecate.git#b3562c2798507869edb767da869cd7b85487726d`,
  `https://github.com/TooTallNate/util-deprecate/tarball/b3562c2798507869edb767da869cd7b85487726d`,
  `git://github.com/TooTallNate/util-deprecate.git#v1.0.1`,
  `git+ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1`,
  `ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1`,
  `git+ssh://git@github.com:yarnpkg/berry.git#v2.1.1`,
  `ssh://git@github.com:yarnpkg/berry.git#v2.1.1`,
  `git+https://github.com/TooTallNate/util-deprecate#v1.0.1`,
  `git+https://github.com/TooTallNate/util-deprecate.git#v1.0.1`,
];

const INVALID_PATTERNS = [
  `./.`,
  `../..`,
];

describe(`gitUtils`, () => {
  for (const pattern of VALID_PATTERNS) {
    it(`should detect ${pattern} as a valid Git url`, () => {
      expect(gitUtils.isGitUrl(pattern)).toEqual(true);
    });
  }

  for (const pattern of INVALID_PATTERNS) {
    it(`shouldn't detect ${pattern} as a valid Git url`, () => {
      expect(gitUtils.isGitUrl(pattern)).toEqual(false);
    });
  }

  for (const pattern of VALID_PATTERNS) {
    it(`should properly normalize ${pattern} ({ git: false })`, () => {
      expect(gitUtils.normalizeRepoUrl(pattern)).toMatchSnapshot();
    });
    it(`should properly normalize ${pattern} ({ git: true })`, () => {
      expect(gitUtils.normalizeRepoUrl(pattern, {git: true})).toMatchSnapshot();
    });
  }

  for (const pattern of VALID_PATTERNS) {
    it(`should properly split ${pattern}`, () => {
      expect(gitUtils.splitRepoUrl(pattern)).toMatchSnapshot();
    });
  }
});
