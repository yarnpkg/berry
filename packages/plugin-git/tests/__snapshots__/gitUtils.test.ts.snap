// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js ({ git: false }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js ({ git: true }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#commit:hash ({ git: false }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#commit:hash"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#commit:hash ({ git: true }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#commit:hash"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#commit=hash ({ git: false }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#commit=hash"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#commit=hash ({ git: true }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#commit=hash"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#commit=hash&workspace=foo ({ git: false }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#commit=hash&workspace=foo"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#commit=hash&workspace=foo ({ git: true }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#commit=hash&workspace=foo"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#hash ({ git: false }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#hash"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#hash ({ git: true }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#hash"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#tag=hello ({ git: false }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#tag=hello"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#tag=hello ({ git: true }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#tag=hello"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#workspace=foo ({ git: false }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#workspace=foo"`;

exports[`gitUtils should properly normalize GitHubOrg/foo-bar.js#workspace=foo ({ git: true }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#workspace=foo"`;

exports[`gitUtils should properly normalize GitHubOrg/foo2bar.js ({ git: false }) 1`] = `"https://github.com/GitHubOrg/foo2bar.js.git"`;

exports[`gitUtils should properly normalize GitHubOrg/foo2bar.js ({ git: true }) 1`] = `"https://github.com/GitHubOrg/foo2bar.js.git"`;

exports[`gitUtils should properly normalize git+https://github.com/TooTallNate/util-deprecate#v1.0.1 ({ git: false }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#v1.0.1"`;

exports[`gitUtils should properly normalize git+https://github.com/TooTallNate/util-deprecate#v1.0.1 ({ git: true }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#v1.0.1"`;

exports[`gitUtils should properly normalize git+https://github.com/TooTallNate/util-deprecate.git#v1.0.1 ({ git: false }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#v1.0.1"`;

exports[`gitUtils should properly normalize git+https://github.com/TooTallNate/util-deprecate.git#v1.0.1 ({ git: true }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#v1.0.1"`;

exports[`gitUtils should properly normalize git+ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1 ({ git: false }) 1`] = `"git+ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1"`;

exports[`gitUtils should properly normalize git+ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1 ({ git: true }) 1`] = `"ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1"`;

exports[`gitUtils should properly normalize git+ssh://git@github.com:yarnpkg/berry.git#v2.1.1 ({ git: false }) 1`] = `"git+ssh://git@github.com:yarnpkg/berry.git#v2.1.1"`;

exports[`gitUtils should properly normalize git+ssh://git@github.com:yarnpkg/berry.git#v2.1.1 ({ git: true }) 1`] = `"git@github.com:yarnpkg/berry.git#v2.1.1"`;

exports[`gitUtils should properly normalize git://github.com/TooTallNate/util-deprecate.git#v1.0.1 ({ git: false }) 1`] = `"git://github.com/TooTallNate/util-deprecate.git#v1.0.1"`;

exports[`gitUtils should properly normalize git://github.com/TooTallNate/util-deprecate.git#v1.0.1 ({ git: true }) 1`] = `"git://github.com/TooTallNate/util-deprecate.git#v1.0.1"`;

exports[`gitUtils should properly normalize github:GitHubOrg/foo-bar.js ({ git: false }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git"`;

exports[`gitUtils should properly normalize github:GitHubOrg/foo-bar.js ({ git: true }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git"`;

exports[`gitUtils should properly normalize github:GitHubOrg/foo-bar.js#hash ({ git: false }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#hash"`;

exports[`gitUtils should properly normalize github:GitHubOrg/foo-bar.js#hash ({ git: true }) 1`] = `"https://github.com/GitHubOrg/foo-bar.js.git#hash"`;

exports[`gitUtils should properly normalize https://github.com/TooTallNate/util-deprecate.git#b3562c2798507869edb767da869cd7b85487726d ({ git: false }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#b3562c2798507869edb767da869cd7b85487726d"`;

exports[`gitUtils should properly normalize https://github.com/TooTallNate/util-deprecate.git#b3562c2798507869edb767da869cd7b85487726d ({ git: true }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#b3562c2798507869edb767da869cd7b85487726d"`;

exports[`gitUtils should properly normalize https://github.com/TooTallNate/util-deprecate.git#master ({ git: false }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#master"`;

exports[`gitUtils should properly normalize https://github.com/TooTallNate/util-deprecate.git#master ({ git: true }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#master"`;

exports[`gitUtils should properly normalize https://github.com/TooTallNate/util-deprecate.git#semver:v1.0.0 ({ git: false }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#semver:v1.0.0"`;

exports[`gitUtils should properly normalize https://github.com/TooTallNate/util-deprecate.git#semver:v1.0.0 ({ git: true }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#semver:v1.0.0"`;

exports[`gitUtils should properly normalize https://github.com/TooTallNate/util-deprecate.git#v1.0.0 ({ git: false }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#v1.0.0"`;

exports[`gitUtils should properly normalize https://github.com/TooTallNate/util-deprecate.git#v1.0.0 ({ git: true }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#v1.0.0"`;

exports[`gitUtils should properly normalize https://github.com/TooTallNate/util-deprecate/tarball/b3562c2798507869edb767da869cd7b85487726d ({ git: false }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#b3562c2798507869edb767da869cd7b85487726d"`;

exports[`gitUtils should properly normalize https://github.com/TooTallNate/util-deprecate/tarball/b3562c2798507869edb767da869cd7b85487726d ({ git: true }) 1`] = `"https://github.com/TooTallNate/util-deprecate.git#b3562c2798507869edb767da869cd7b85487726d"`;

exports[`gitUtils should properly normalize ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1 ({ git: false }) 1`] = `"ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1"`;

exports[`gitUtils should properly normalize ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1 ({ git: true }) 1`] = `"ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1"`;

exports[`gitUtils should properly normalize ssh://git@github.com:yarnpkg/berry.git#v2.1.1 ({ git: false }) 1`] = `"ssh://git@github.com:yarnpkg/berry.git#v2.1.1"`;

exports[`gitUtils should properly normalize ssh://git@github.com:yarnpkg/berry.git#v2.1.1 ({ git: true }) 1`] = `"git@github.com:yarnpkg/berry.git#v2.1.1"`;

exports[`gitUtils should properly split GitHubOrg/foo-bar.js 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/GitHubOrg/foo-bar.js.git",
  "treeish": Object {
    "protocol": "head",
    "request": "HEAD",
  },
}
`;

exports[`gitUtils should properly split GitHubOrg/foo-bar.js#commit:hash 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/GitHubOrg/foo-bar.js.git",
  "treeish": Object {
    "protocol": "commit",
    "request": "hash",
  },
}
`;

exports[`gitUtils should properly split GitHubOrg/foo-bar.js#commit=hash 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/GitHubOrg/foo-bar.js.git",
  "treeish": Object {
    "protocol": "commit",
    "request": "hash",
  },
}
`;

exports[`gitUtils should properly split GitHubOrg/foo-bar.js#commit=hash&workspace=foo 1`] = `
Object {
  "extra": Object {
    "workspace": "foo",
  },
  "repo": "https://github.com/GitHubOrg/foo-bar.js.git",
  "treeish": Object {
    "protocol": "commit",
    "request": "hash",
  },
}
`;

exports[`gitUtils should properly split GitHubOrg/foo-bar.js#hash 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/GitHubOrg/foo-bar.js.git",
  "treeish": Object {
    "protocol": null,
    "request": "hash",
  },
}
`;

exports[`gitUtils should properly split GitHubOrg/foo-bar.js#tag=hello 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/GitHubOrg/foo-bar.js.git",
  "treeish": Object {
    "protocol": "tag",
    "request": "hello",
  },
}
`;

exports[`gitUtils should properly split GitHubOrg/foo-bar.js#workspace=foo 1`] = `
Object {
  "extra": Object {
    "workspace": "foo",
  },
  "repo": "https://github.com/GitHubOrg/foo-bar.js.git",
  "treeish": Object {
    "protocol": "head",
    "request": "HEAD",
  },
}
`;

exports[`gitUtils should properly split GitHubOrg/foo2bar.js 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/GitHubOrg/foo2bar.js.git",
  "treeish": Object {
    "protocol": "head",
    "request": "HEAD",
  },
}
`;

exports[`gitUtils should properly split git+https://github.com/TooTallNate/util-deprecate#v1.0.1 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/TooTallNate/util-deprecate.git",
  "treeish": Object {
    "protocol": null,
    "request": "v1.0.1",
  },
}
`;

exports[`gitUtils should properly split git+https://github.com/TooTallNate/util-deprecate.git#v1.0.1 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/TooTallNate/util-deprecate.git",
  "treeish": Object {
    "protocol": null,
    "request": "v1.0.1",
  },
}
`;

exports[`gitUtils should properly split git+ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1 1`] = `
Object {
  "extra": Object {},
  "repo": "git+ssh://git@github.com/TooTallNate/util-deprecate.git",
  "treeish": Object {
    "protocol": null,
    "request": "v1.0.1",
  },
}
`;

exports[`gitUtils should properly split git+ssh://git@github.com:yarnpkg/berry.git#v2.1.1 1`] = `
Object {
  "extra": Object {},
  "repo": "git+ssh://git@github.com:yarnpkg/berry.git",
  "treeish": Object {
    "protocol": null,
    "request": "v2.1.1",
  },
}
`;

exports[`gitUtils should properly split git://github.com/TooTallNate/util-deprecate.git#v1.0.1 1`] = `
Object {
  "extra": Object {},
  "repo": "git://github.com/TooTallNate/util-deprecate.git",
  "treeish": Object {
    "protocol": null,
    "request": "v1.0.1",
  },
}
`;

exports[`gitUtils should properly split github:GitHubOrg/foo-bar.js 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/GitHubOrg/foo-bar.js.git",
  "treeish": Object {
    "protocol": "head",
    "request": "HEAD",
  },
}
`;

exports[`gitUtils should properly split github:GitHubOrg/foo-bar.js#hash 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/GitHubOrg/foo-bar.js.git",
  "treeish": Object {
    "protocol": null,
    "request": "hash",
  },
}
`;

exports[`gitUtils should properly split https://github.com/TooTallNate/util-deprecate.git#b3562c2798507869edb767da869cd7b85487726d 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/TooTallNate/util-deprecate.git",
  "treeish": Object {
    "protocol": null,
    "request": "b3562c2798507869edb767da869cd7b85487726d",
  },
}
`;

exports[`gitUtils should properly split https://github.com/TooTallNate/util-deprecate.git#master 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/TooTallNate/util-deprecate.git",
  "treeish": Object {
    "protocol": null,
    "request": "master",
  },
}
`;

exports[`gitUtils should properly split https://github.com/TooTallNate/util-deprecate.git#semver:v1.0.0 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/TooTallNate/util-deprecate.git",
  "treeish": Object {
    "protocol": "semver",
    "request": "v1.0.0",
  },
}
`;

exports[`gitUtils should properly split https://github.com/TooTallNate/util-deprecate.git#v1.0.0 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/TooTallNate/util-deprecate.git",
  "treeish": Object {
    "protocol": null,
    "request": "v1.0.0",
  },
}
`;

exports[`gitUtils should properly split https://github.com/TooTallNate/util-deprecate/tarball/b3562c2798507869edb767da869cd7b85487726d 1`] = `
Object {
  "extra": Object {},
  "repo": "https://github.com/TooTallNate/util-deprecate.git",
  "treeish": Object {
    "protocol": null,
    "request": "b3562c2798507869edb767da869cd7b85487726d",
  },
}
`;

exports[`gitUtils should properly split ssh://git@github.com/TooTallNate/util-deprecate.git#v1.0.1 1`] = `
Object {
  "extra": Object {},
  "repo": "ssh://git@github.com/TooTallNate/util-deprecate.git",
  "treeish": Object {
    "protocol": null,
    "request": "v1.0.1",
  },
}
`;

exports[`gitUtils should properly split ssh://git@github.com:yarnpkg/berry.git#v2.1.1 1`] = `
Object {
  "extra": Object {},
  "repo": "ssh://git@github.com:yarnpkg/berry.git",
  "treeish": Object {
    "protocol": null,
    "request": "v2.1.1",
  },
}
`;
